stix2/base.py

```diff
@@ -347,6 +360,61 @@ def _check_property(self, prop_name, prop, kwargs):
             for ref in kwargs[prop_name]:
                 self._check_ref(ref, prop, prop_name)
 
+    def _generate_id(self, kwargs):
+        required_prefix = self._type + "--"
+        namespace = uuid.UUID("00abedb4-aa42-466c-9c01-fed23315a9b7")
+
+        try:
+            properties_to_use = self._id_contributing_properties
+            if properties_to_use:
+                streamlined_object = {}
+                if "hashes" in kwargs and "hashes" in properties_to_use:
+                    possible_hash = self._choose_one_hash(kwargs["hashes"])
+                    if possible_hash:
+                        streamlined_object["hashes"] = possible_hash
+                for key in kwargs.keys():
+                    if key in properties_to_use and key != "hashes":
+                        if type(kwargs[key]) is dict:
+                            for otherKey in kwargs[key]:
+                                if isinstance(kwargs[key][otherKey], _STIXBase):
+                                    streamlined_object[key] = self._embed_obj_to_json(kwargs[key][otherKey])
+                                else:
+                                    streamlined_object[key] = kwargs[key]
+                        else:
+                            if isinstance(kwargs[key], _STIXBase):
+                                streamlined_object[key] = self._embed_obj_to_json(kwargs[key])
+                            else:
+                                streamlined_object[key] = kwargs[key]
+
+                if streamlined_object:
+                    data = canonicalize(str(streamlined_object), utf8=False)
+                    return required_prefix + str(uuid.uuid5(namespace, str(data)))
+            return None
+        except AttributeError:
+            # We ideally end up here if handling a 2.0 SCO
+            return None
+
+    def _choose_one_hash(self, hash_dict):
+        if "MD5" in hash_dict:
+            return {"MD5": hash_dict["MD5"]}
+        elif "SHA-1" in hash_dict:
+            return {"SHA-1": hash_dict["SHA-1"]}
+        elif "SHA-256" in hash_dict:
+            return {"SHA-256": hash_dict["SHA-256"]}
+        elif "SHA-512" in hash_dict:
+            return {"SHA-512": hash_dict["SHA-512"]}
+        else:
+            # Cheesy way to pick the first item in the dictionary, since its not indexable
+            for (k, v) in hash_dict.items():
+                break
+            return {k: v}
```

Comment:
This will fail if the dict is empty (k and v will not have been assigned values).  I'd prefer something more bulletproof (even if through some logic elsewhere we don't think that will ever happen ;))

Here's a better way to get an arbitrary key from a dict:
```python
k = next(iter(hash_dict), None)
```
This will result in `None` if the dict is empty (of course, we assume None itself wasn't a key!).  You could also get a key/value tuple if you wanted:
```python
kv = next(iter(hash_dict.items()), None)
```
Just be careful to check `kv` for None before trying to get the key and value from the tuple, in case the dict was empty.